### PR TITLE
Implement mergeSort() function

### PR DESCRIPTION
**Pull request by yhabteab**
resoves #12

### CODE REVIEW

File: sort.php [hunk 1]
```diff
@@ -48,6 +48,41 @@ function quickSort($arr, $funcCallback)
...
+        return $arr;
+    }
+
+    $mergedArray = array();
+
+    $middle = round(sizeof($arr) / 2, 0, PHP_ROUND_HALF_DOWN);
+    $leftSide = mergeSort(array_slice($arr, 0, $middle), $funcCallback);
+    $rightSide = mergeSort(array_slice($arr, $middle, sizeof($arr)), $funcCallback);
+
+    while ($funcCallback(0, sizeof($leftSide)) && $funcCallback(0, sizeof($rightSide))) {
```
Comment (Al2Klimov) [inline sort.php]: Why are you using $funcCallback for values not from your input?

File: sort.php [hunk 2]
```diff
@@ -48,6 +48,41 @@ function quickSort($arr, $funcCallback)
...
+        if ($funcCallback($leftSide[0], $rightSide[0])) {
+            $mergedArray[] = $leftSide[0];
+            $leftSide = array_slice($leftSide, 1);
+        } else {
+            $mergedArray[] = $rightSide[0];
+            $rightSide = array_slice($rightSide, 1);
+        }
+    }
+
+    while ($funcCallback(0, sizeof($leftSide))) {
```
Comment (Al2Klimov) [inline sort.php]: Same here.

File: sort.php [hunk 3]
```diff
@@ -48,6 +48,41 @@ function quickSort($arr, $funcCallback)
...
+        $mergedArray[] = $leftSide[0];
+        $leftSide = array_slice($leftSide, 1);
+    }
+
+    while ($funcCallback(0, sizeof($rightSide))) {
```
Comment (Al2Klimov) [inline sort.php]: And here.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: sort.php [hunk 4]
```diff
@@ -48,6 +48,41 @@ function quickSort($arr, $funcCallback)
...
+        } else {
+            $mergedArray[] = $rightSide[0];
+            $rightSide = array_slice($rightSide, 1);
+        }
+    }
+
+    while (0 < sizeof($leftSide)) {
+        $mergedArray[] = $leftSide[0];
+        $leftSide = array_slice($leftSide, 1);
+    }
```
Comment (Al2Klimov) [inline sort.php]: Any way to shorten this, maybe to a one-liner?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: sort.php (hunk 4 above)
Comment (Al2Klimov) [reply sort.php]: I meant the whole while.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
* [ ] `git reset --soft HEAD~$(git rev-list --count $(git merge-base HEAD master)..HEAD^)`
* [ ] `git commit --amend --no-edit`
* [ ] `git push -f origin feature/sort-#12`

**Comment by yhabteab** [comment]
##Testprotocol #12

Test protocol for ``--mergesort`` parameter.

The content of the in.txt file looks like this.
```
13
8
12
6
2
10
1
7
0
```
Commands with the ``--mergesort`` and ``--numeric-sort`` parameter.
``php sort.php --mergesort -n in.txt`` AND ``php sort.php --mergesort -n < in.txt``  output:
```
0                                             0
1                                             1
2                                             2
6                                             6
7                                             7
8                                             8
10                                            10
12                                            12
13                                            13
```
and without ``--numeric-sort`` parameter the output looks like this..
``php sort.php --mergesort in.txt`` AND ``php sort.php --mergesort < in.txt``
```
0                                       0
1                                       1
10                                      10
12                                      12
13                                      13
2                                       2
6                                       6
7                                       7
8                                       8
```

See you next time!

:+1:

**Review by Al2Klimov — APPROVED** [review]

**Pull request merged**